File: driver_app/lib/core/realtime/realtime_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

import '../config/api_config.dart';
import '../storage/secure_token_storage.dart';

typedef RealtimeEventHandler = void Function(Map<String, dynamic> event);

class RealtimeClient {
  RealtimeClient(this._storage);

  final SecureTokenStorage _storage;
  WebSocketChannel? _channel;
  StreamSubscription<dynamic>? _subscription;
  Timer? _reconnectTimer;
  Timer? _pingTimer;
  bool _shouldReconnect = false;
  int _reconnectAttempts = 0;
  final _handlers = <RealtimeEventHandler>{};

  void addListener(RealtimeEventHandler handler) => _handlers.add(handler);

  void removeListener(RealtimeEventHandler handler) => _handlers.remove(handler);

  Future<void> connect() async {
    final tokens = await _storage.readTokens();
    if (tokens == null) return;
    _shouldReconnect = true;
    await _open(tokens.accessToken);
  }

  Future<void> disconnect() async {
    _shouldReconnect = false;
    _reconnectTimer?.cancel();
    _pingTimer?.cancel();
    await _subscription?.cancel();
    await _channel?.sink.close();
    _channel = null;
  }

  void subscribeOrder(String orderId) {
    _send({'action': 'subscribe', 'channel': 'order', 'order_id': orderId});
  }

  void updatePresence({required String status, String? activeOrderId}) {
    _send({
      'action': 'presence',
      'status': status,
      if (activeOrderId != null) 'active_order_id': activeOrderId,
    });
  }

  Future<void> _open(String token) async {
    await _subscription?.cancel();
    await _channel?.sink.close();

    final base = ApiConfig.baseUrl.replaceFirst('http', 'ws');
    final uri = Uri.parse('$base/ws?token=${Uri.encodeComponent(token)}');
    _channel = WebSocketChannel.connect(uri);
    _subscription = _channel!.stream.listen(
      _onMessage,
      onDone: _scheduleReconnect,
      onError: (_) => _scheduleReconnect,
    );
    _reconnectAttempts = 0;
    _pingTimer?.cancel();
    _pingTimer = Timer.periodic(const Duration(seconds: 25), (_) {
      _send({'action': 'ping'});
    });
  }

  void _onMessage(dynamic message) {
    final decoded = jsonDecode(message as String) as Map<String, dynamic>;
    if (decoded['type'] == 'BATCH') {
      final events = decoded['events'] as List<dynamic>? ?? [];
      for (final item in events) {
        _dispatch(item as Map<String, dynamic>);
      }
      return;
    }
    _dispatch(decoded);
  }

  void _dispatch(Map<String, dynamic> event) {
    if (event['type'] == 'HEARTBEAT') return;
    for (final handler in _handlers) {
      handler(event);
    }
  }

  void _send(Map<String, dynamic> payload) {
    final channel = _channel;
    if (channel == null) return;
    channel.sink.add(jsonEncode(payload));
  }

  void _scheduleReconnect() {
    if (!_shouldReconnect) return;
    _reconnectTimer?.cancel();
    final delay = Duration(seconds: (1 << _reconnectAttempts.clamp(0, 5)));
    _reconnectAttempts += 1;
    _reconnectTimer = Timer(delay, () async {
      final tokens = await _storage.readTokens();
      if (tokens != null) {
        await _open(tokens.accessToken);
      }
    });
  }
}

File: frontend/src/services/realtimeClient.ts
import type { QueryClient } from "@tanstack/react-query";

export type RealtimeSeverity = "info" | "success" | "warning" | "error";

export interface RealtimeEvent {
  id: string;
  type: string;
  company_id: string;
  channel: string;
  payload: Record<string, unknown>;
  severity?: RealtimeSeverity;
  created_at: string;
}

export type ConnectionState = "connecting" | "connected" | "disconnected" | "error";

type EventHandler = (event: RealtimeEvent) => void;
type StateHandler = (state: ConnectionState) => void;

function wsBaseUrl(): string {
  const protocol = window.location.protocol === "https:" ? "wss:" : "ws:";
  return `${protocol}//${window.location.host}/api/v1/ws`;
}

export class RealtimeClient {
  private socket: WebSocket | null = null;
  private handlers = new Set<EventHandler>();
  private stateHandlers = new Set<StateHandler>();
  private reconnectAttempts = 0;
  private reconnectTimer: number | null = null;
  private pingTimer: number | null = null;
  private token: string | null = null;
  private shouldReconnect = true;

  connect(accessToken: string): void {
    this.token = accessToken;
    this.shouldReconnect = true;
    this.openSocket();
  }

  disconnect(): void {
    this.shouldReconnect = false;
    this.clearTimers();
    this.socket?.close();
    this.socket = null;
    this.setState("disconnected");
  }

  subscribe(handler: EventHandler): () => void {
    this.handlers.add(handler);
    return () => this.handlers.delete(handler);
  }

  onStateChange(handler: StateHandler): () => void {
    this.stateHandlers.add(handler);
    return () => this.stateHandlers.delete(handler);
  }

  subscribeOrder(orderId: string): void {
    this.send({ action: "subscribe", channel: "order", order_id: orderId });
  }

  private openSocket(): void {
    if (!this.token) return;
    this.clearTimers();
    this.setState("connecting");
    const socket = new WebSocket(`${wsBaseUrl()}?token=${encodeURIComponent(this.token)}`);
    this.socket = socket;

    socket.onopen = () => {
      this.reconnectAttempts = 0;
      this.setState("connected");
      this.pingTimer = window.setInterval(() => {
        this.send({ action: "ping" });
      }, 25_000);
    };

    socket.onmessage = (message) => {
      try {
        const payload = JSON.parse(message.data) as RealtimeEvent | { type: "BATCH"; events: RealtimeEvent[] };
        if (payload.type === "BATCH" && Array.isArray(payload.events)) {
          payload.events.forEach((event) => this.emit(event));
          return;
        }
        this.emit(payload as RealtimeEvent);
      } catch {
        // Ignore malformed frames.
      }
    };

    socket.onerror = () => this.setState("error");

    socket.onclose = () => {
      this.clearTimers();
      this.setState("disconnected");
      if (this.shouldReconnect) {
        const delay = Math.min(30_000, 1000 * 2 ** this.reconnectAttempts);
        this.reconnectAttempts += 1;
        this.reconnectTimer = window.setTimeout(() => this.openSocket(), delay);
      }
    };
  }

  private emit(event: RealtimeEvent): void {
    if (event.type === "HEARTBEAT") return;
    this.handlers.forEach((handler) => handler(event));
  }

  private send(payload: Record<string, unknown>): void {
    if (this.socket?.readyState === WebSocket.OPEN) {
      this.socket.send(JSON.stringify(payload));
    }
  }

  private setState(state: ConnectionState): void {
    this.stateHandlers.forEach((handler) => handler(state));
  }

  private clearTimers(): void {
    if (this.reconnectTimer !== null) {
      window.clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.pingTimer !== null) {
      window.clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }
}

export const realtimeClient = new RealtimeClient();

const ORDER_EVENT_PREFIXES = ["ORDER_", "LOADING_", "TRANSIT_", "DELIVERY_", "DRIVER_ARRIVED", "ARRIVED_"];

export function applyRealtimeEvent(queryClient: QueryClient, event: RealtimeEvent): void {
  const { type, payload } = event;
  const orderId = typeof payload.order_id === "string" ? payload.order_id : null;

  if (type === "NOTIFICATION_CREATED") {
    queryClient.invalidateQueries({ queryKey: ["notifications"] });
    return;
  }

  if (type === "TIMELINE_ENTRY" && orderId) {
    queryClient.invalidateQueries({ queryKey: ["orders", orderId, "timeline"] });
  }

  if (type === "PRESENCE_UPDATED" || type === "DRIVER_CONNECTED" || type === "DRIVER_DISCONNECTED") {
    queryClient.invalidateQueries({ queryKey: ["presence"] });
    queryClient.invalidateQueries({ queryKey: ["dashboard"] });
  }

  if (ORDER_EVENT_PREFIXES.some((prefix) => type.startsWith(prefix)) || type.includes("VIN") || type.includes("PHOTO") || type.includes("DAMAGE") || type.includes("DOCUMENT") || type.includes("CMR")) {
    queryClient.invalidateQueries({ queryKey: ["dashboard"] });
    queryClient.invalidateQueries({ queryKey: ["orders"] });
    if (orderId) {
      queryClient.invalidateQueries({ queryKey: ["orders", orderId] });
      queryClient.invalidateQueries({ queryKey: ["orders", orderId, "timeline"] });
      queryClient.invalidateQueries({ queryKey: ["orders", orderId, "checklist"] });
      queryClient.invalidateQueries({ queryKey: ["orders", orderId, "documents"] });
    }
  }
}
